plog-client/src/main/java/com/airbnb/plog/client/PlogClient.java

```diff
@@ -0,0 +1,156 @@
+
+    openSocket();
+    try {
+      address = InetAddress.getByName(host);
+    } catch (UnknownHostException e) {
+      log.error("Unknown address {}", host, e);
+      throw Throwables.propagate(e);
+    }
+    this.port = port;
+    this.chunkSize = chunkSize;
+    this.lastMessageId = new AtomicInteger(1);
+  }
+
+  /**
+   * Send the message to Plog server.
+   */
+  public void send(String message) {
+    // ISO-8859-1 is ASCII-8bit. It's equivalent to ruby's BINARY encoding.
+    byte[] messageBytes = message.getBytes(Charset.forName("ISO-8859-1"));
+    int messageId = lastMessageId.getAndIncrement();
+    lastMessageId.compareAndSet(Integer.MAX_VALUE, 1);
+    int messageLength = messageBytes.length;
+    byte[] checksum = computeChecksum(messageBytes);
+    List<byte[]> chunks = chunkMessage(messageBytes, chunkSize);
+    int count = chunks.size();
+
+    log.debug("Plog: sending {}; {} chunk(s)", messageId, count);
+    for (int i = 0; i < count; i++) {
+      sendToSocket(MultipartMessage.encode(messageId,
+          messageLength,
+          checksum,
+          chunkSize,
+          count,
+          i,
+          chunks.get(i)));
+    }
+  }
+
+  @VisibleForTesting
+  static byte[] computeChecksum(byte[] messageBytes) {
+    // Checksum in little-endian.
+    byte[] checksum = Hashing.murmur3_32().hashBytes(messageBytes).asBytes();
+    // Reverse checksum bytes to make it big-endian.
+    byte temp;
+    int start = 0, end = 3;
+    while (start < end) {
+      temp = checksum[start];
+      checksum[start] = checksum[end];
+      checksum[end] = temp;
+      start++;
+      end--;
+    }
+    return checksum;
+  }
+
+  @VisibleForTesting
+  static List<byte[]> chunkMessage(byte[] messageBytes, int size) {
+    final List<byte[]> chunks = new ArrayList<byte[]>();
+    int startIndex = 0;
+    while (startIndex + size < messageBytes.length) {
+      chunks.add(Arrays.copyOfRange(messageBytes, startIndex, startIndex + size));
+      startIndex += size;
+    }
+    // If there's some remaining bytes,
+    // copy them up to the end of messageBytes.
+    if (startIndex < messageBytes.length) {
+      chunks.add(Arrays.copyOfRange(messageBytes, startIndex, messageBytes.length));
+    }
+    return chunks;
+  }
+
+  private void openSocket() {
+    try {
+      socket = new DatagramSocket();
+    } catch (SocketException e) {
+      log.error("Cannot open socket", e);
+      throw Throwables.propagate(e);
+    }
+  }
+
+  private void sendToSocket(byte[] chunk) {
+    DatagramPacket packet = new DatagramPacket(chunk, chunk.length, address, port);
+    try {
+      log.trace("Sending {} to UDP port {}", chunk, port);
+      socket.send(packet);
+    } catch (IOException e) {
+      log.error("Error sending packet!", e);
+      socket.close();
+      openSocket();
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    socket.close();
```

Comment:
`socket` might be `null` if an exception was thrown in `sendToSocket` `catch` clause.